stm-doc/index.mld

```diff
@@ -0,0 +1,451 @@
+
+--- Failure --------------------------------------------------------------------
+
+Test STM sequential tests failed (4 shrink steps):
+
+[(Add 3591134320860609976); Length]
+
++++ Messages ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+
+Messages for test STM sequential tests:
+
+  Results incompatible with model
+
+   (Add 3591134320860609976) : ()
+   Length : 0
+================================================================================
+failure (1 tests failed, 0 tests errored, ran 1 tests)
+]}
+
+Here, the test is failing.
+
+In the case of a failing test, [qcheck-stm] gives the user the counter example
+it has found, after some shrinking steps, as it is customary in property based
+testing {i à la} QuickCheck. The counter example is a program. It is given
+twice. The first time just as a list of calls. The second time, each call is
+paired with the the result of the computation.
+
+Looking at the output, we see that [Add 3591134320860609976] returns a [unit],
+which is to be expected. But the call to [Length] returns [0] despite the fact
+that it has been ran after the [Add]. We conclude that there is something wrong
+with our implementation of [add].
+
+Indeed, we've forgotten to increment the [length] field.
+
+{2 Second version}
+
+In the second version, we will fix this problem. This is easily done just by
+updating the [length] field when adding an element.
+
+{[
+module Lib ...
+
+  let add a t =
+    if not (mem a t)
+    then begin
+      t.content <- a :: t.content;
+      t.length  <- t.length + 1;
+    end
+
+  ...
+
+end
+]}
+
+Then, the output looks like that:
+
+{[
+$ dune exec ./mutable_set.exe
+random seed: 364286126
+generated error fail pass / total     time test name
+[✓]  100    0    0  100 /  100     0.1s STM sequential tests (generating)
+================================================================================
+success (ran 1 tests)
+]}
+
+We can see that the tests are successful. That means that the real behaviour of
+the one hundred generated programs were consistent with the functional model
+we've given.
+
+{2 Third version -- parallel tests}
+
+We've now tested that, when run sequentially, our library behave accordingly to
+the functional model we've given. The power of [qcheck-stm] comes from the fact
+that we can also test the behaviour of our library when run in a parallel or
+concurrent context.
+
+Let's say we want to test our library when it is used with OCaml domains. We
+just have to instantiate our specification with another functor:
+
+
+{[
+(* same as before *)
+
+module Lib_domain = STM_domain.Make(Lib_spec)
+
+let _ = QCheck_runner.run_tests ~verbose:true [Lib_domain.agree_test_par ~count:100 ~name:"STM parallel tests"]
+]}
+
+In the dune file, we should also update the dependency: [qcheck-stm.sequential]
+and [qcheck-stm.domain] are two different libraries. In the same manner, if you
+want to test your library with [Thread]s, you'll depends on
```

Comment:
`you'll depends on` -> `you will need a dependency on`

I would perhaps prefer not to advertise the `Thread`-mode in the tutorial, since it is still preliminary and instead focus on the `Domain` mode that works well.

Reply:
Finally, I've removed the dune file from the tutorial (it's mostly noise). And I agree about not advertising the `Thread`-mode in the tutorial.